### PR TITLE
Add API rz_flag_get_name_delta

### PR DESCRIPTION
**Pull request by p371jah**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Review by ret2libc — APPROVED** [review]
<3

File: librz/flag/flag.c [hunk 1]
```diff
@@ -637,6 +637,22 @@ RZ_API RzFlagItem *rz_flag_get_at(RzFlag *f, ut64 off, bool closest) {
 	return nice ? evalFlag(f, nice) : NULL;
 }
 
+/**
+ * \brief Returns name of the closest flag with offset
+ */
+RZ_API RZ_OWN char *rz_flag_get_name_delta(RZ_NONNULL RzFlag *f, ut64 addr) {
+	rz_return_val_if_fail(f, NULL);
+	RzFlagItem *item = rz_flag_get_at(f, addr, true);
+	if (item) {
```
Comment (XVilka) [inline librz/flag/flag.c]: This condition could be reversed.

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
Please fix the tests:
```
 [XX] db/formats/mach0/objc aao hello-objc selrefs
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'e analysis.jmp.cref=true
aao
axl~?
axl~selref
' bins/mach0/hello-objc
-- stdout
--- expected
+++ actual
@@ --1,10 +-1,10 @@
 92
-              section.0.__TEXT.__text+20 0x100000db4 ->      DATA -> 0x100001150 section.14.__DATA.__objc_selrefs
-                                 main+15 0x100000e5f ->      DATA -> 0x100001158 section.14.__DATA.__objc_selrefs+8
-                                 main+45 0x100000e7d ->      DATA -> 0x100001160 section.14.__DATA.__objc_selrefs+16
-                                 main+64 0x100000e90 ->      DATA -> 0x100001168 section.14.__DATA.__objc_selrefs+24
-                                 main+86 0x100000ea6 ->      DATA -> 0x100001158 section.14.__DATA.__objc_selrefs+8
-                                main+116 0x100000ec4 ->      DATA -> 0x100001160 section.14.__DATA.__objc_selrefs+16
-                                main+135 0x100000ed7 ->      DATA -> 0x100001168 section.14.__DATA.__objc_selrefs+24
+            section.0.__TEXT.__text + 20 0x100000db4 ->      DATA -> 0x100001150 section.14.__DATA.__objc_selrefs
+                               main + 15 0x100000e5f ->      DATA -> 0x100001158 section.14.__DATA.__objc_selrefs + 8
+                               main + 45 0x100000e7d ->      DATA -> 0x100001160 section.14.__DATA.__objc_selrefs + 16
+                               main + 64 0x100000e90 ->      DATA -> 0x100001168 section.14.__DATA.__objc_selrefs + 24
+                               main + 86 0x100000ea6 ->      DATA -> 0x100001158 section.14.__DATA.__objc_selrefs + 8
+                              main + 116 0x100000ec4 ->      DATA -> 0x100001160 section.14.__DATA.__objc_selrefs + 16
+                              main + 135 0x100000ed7 ->      DATA -> 0x100001168 section.14.__DATA.__objc_selrefs + 24

  --
 __auth_got (auth, bind)
-0x100004000 0xc009000000000000 
-0x100004008 0xc009000000000005 
-0x100004010 0xc009000000000006 
-0x100004018 0xc009000000000007 
-0x100004020 0xc009000000000008 
+0x100004000 0xc009000000000000 segment.LINKEDIT + 13837591325777543168
+0x100004008 0xc009000000000005 segment.LINKEDIT + 13837591325777543173
+0x100004010 0xc009000000000006 segment.LINKEDIT + 13837591325777543174
+0x100004018 0xc009000000000007 segment.LINKEDIT + 13837591325777543175
+0x100004020 0xc009000000000008 segment.LINKEDIT + 13837591325777543176
 --
 cstring (!auth, !bind)
 - offset -    0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
```

**Comment by p371jah** [comment]
`rz_flag_get_at(core->flags, addr, true)` returns `flag` such that `flag->offset <= addr` but it doesn't mean that `flag->offset <= addr <= flag->(offset + flag->size)`. Now `px` is spammy because previously there was a hackish check.

File: librz/flag/flag.c [hunk 2]
```diff
@@ -637,6 +637,22 @@ RZ_API RzFlagItem *rz_flag_get_at(RzFlag *f, ut64 off, bool closest) {
...
+ * \brief Returns name of the closest flag with offset
+ */
+RZ_API RZ_OWN char *rz_flag_get_name_delta(RZ_NONNULL RzFlag *f, ut64 addr) {
+	rz_return_val_if_fail(f, NULL);
+	RzFlagItem *item = rz_flag_get_at(f, addr, true);
+	if (!item) {
+		return NULL;
+	}
+	if (item->offset == addr) {
+		return strdup(item->name);
```
Comment (XVilka) [inline librz/flag/flag.c:668]: Shouldn't the `realname | name` logic be used here too, as below?

**Comment by ret2libc** [comment]
@pelijah I'm ok with adjusting tests if we think new behaviour is better. Overall, I think this is a very good change so there should be a good way to move this forward and I wouldn't block just for one single command that can be adjusted a bit.

**Comment by p371jah** [comment]
But the thing is I don't think that new behaviour is better.

**Comment by ret2libc** [comment]
> `rz_flag_get_at(core->flags, addr, true)` returns `flag` such that `flag->offset <= addr` but it doesn't mean that `flag->offset <= addr <= flag->(offset + flag->size)`. Now `px` is spammy because previously there was a hackish check.

What happens if you actually check within `rz_flag_get_name_delta` that `addr <= flag->offset + flag->size`? Would this break many things? I think it's the only check that makes sense, because 1) if you have a flag at 0 and you request addr=0xcafebabe you don't want to get `flag0 + 0xcafebabe`, unless the flag was spanning that much space. 2) having random checks with random numbers e.g. `delta < 8192` is meh IMO.

File: librz/flag/flag.c [hunk 3]
```diff
@@ -637,6 +637,40 @@ RZ_API RzFlagItem *rz_flag_get_at(RzFlag *f, ut64 off, bool closest) {
...
+	}
+	RzListIter *iter;
+	RzFlagItem *item;
+	rz_list_foreach (flags_at->flags, iter, item) {
+		if ((item->offset + item->size) >= off) {
+			return item;
+		}
+	}
+	return NULL;
+}
```
Comment (p371jah) [inline librz/flag/flag.c:656]: @ret2libc we need something like this.

**Comment by p371jah** [comment]
Meh RzFlag API is a big mess. It was a bad idea to touch it.

**Comment by ret2libc** [comment]
> Meh RzFlag API is a big mess. It was a bad idea to touch it.

Why did you close this? :(

I thought there were just 1 or 2 tests failing, what was the blocker there? I think the PR was very good actually!

**Pull request closed without merging**